File: Chapter 2/Section 2.2/runround/runround.cpp
/*
ID: ten.to.1
TASK: runround
LANG: C++
*/

#include <iostream>
#include <fstream>
#include <string.h>

bool is_unique(std::string digits)
{
	for(int i = 0; i < digits.length(); i++)
	{
		for(int j = i + 1; j < digits.length(); j++)
		{
			if(digits[i] == digits[j])
				return false;
		}
	}
	
	return true;
}

bool is_runround(int n)
{
	std::string digits = std::to_string(n);
	bool touched[digits.length()];
	memset(touched, false, digits.length());
	
	if(!is_unique(digits))
		return false;
	
	for(int i = 0; i < digits.length(); i++)
		if(digits[i] == '0')
			return false;
	
	int moves = 0;
	
	for(int i = 0; i < digits.length(); i++)
	{
		//-48 to convert ASCII value to Decimal value
		moves += digits[moves % digits.length()] - 48;
		touched[moves % digits.length()] = true;
	}
	
	for(int i = 0; i < digits.length(); i++)
		if(!touched[i])
			return false;
	
	if(digits[moves % digits.length()] == digits[0])
		return true;
	else
		return false;
	
}

int main()
{
	std::ifstream file_in("runround.in");
	std::ofstream file_out("runround.out");
	
	int M; file_in >> M;
	
	M++;
	while(!is_runround(M))
		M++;
	
	file_out << M << '\n';
}

File: Chapter 2/Section 2.2/runround/runround.py
"""
ID: ten.to.1
TASK: runround
LANG: PYTHON3
"""

#!/usr/bin/python

file_in  = open("runround.in", "r")
file_out = open("runround.out", "w")

M = int(file_in.read())

def is_unique(digits):
	for n in range(0, len(digits)):
		for j in range(n + 1, len(digits)):
			if(digits[n] == digits[j]):
				return False
	
	return True

def is_runround(n):
	digits  = [int(x) for x in str(n)]
	touched = [False] * len(digits)
	
	if(not is_unique(digits)):
		return False
	
	for n in digits:
		if(n == 0):
			return False
	
	moves = 0
	
	for n in range(0, len(digits)):
		moves += digits[moves % len(digits)]
		touched[moves % len(digits)] = True
	
	for t in touched:
		if(not t):
			return False
	
	if(digits[moves % len(digits)] == digits[0]):
		return True
	else:
		return False

M += 1
while(not is_runround(M)):
	M += 1

file_out.write(str(M) + "\n")

